ocaml/xapi-guard/lib/disk_cache.ml

```diff
@@ -0,0 +1,596 @@
+          Lwt.return_unit
+      | exn ->
+          D.info "%s: error when reading '%s': %s" __FUN path
+            Printexc.(to_string exn) ;
+          Lwt.return_unit
+    in
+
+    Lwt.try_bind
+      (fun () -> read_from ~filename:path)
+      (fun contents ->
+        let* () = retry_push push key contents in
+        unlink_safe path
+      )
+      on_error
+
+  let push_files push files = Lwt_list.iter_s (push_file push) (List.rev files)
+
+  let update_all queue push root =
+    let __FUN = __FUNCTION__ in
+    let* contents = get_latest_and_delete_rest root in
+    let* () = push_files push contents in
+    let@ () = with_lock queue.lock in
+    let* contents = get_latest_and_delete_rest root in
+    let* () =
+      match contents with
+      | [] ->
+          queue.state <- Direct ;
+          D.debug "%s: Cache clean; Going direct" __FUN ;
+          Lwt.return_unit
+      | _ ->
+          Lwt.return_unit
+    in
+    Lwt.return_unit
+
+  let resolve queue push root = function
+    | File file -> (
+        let* () = push_file push file in
+        let@ () = with_lock queue.lock in
+        match queue.state with
+        | Direct | Disengaged ->
+            Lwt.return_unit
+        | Engaged ->
+            let () =
+              if Lwt_bounded_stream.size queue.queue = 0 then
+                queue.state <- Direct
+            in
+            Lwt.return_unit
+      )
+    | Update_all ->
+        update_all queue push root
+    | Wait ->
+        (* Do not busy loop when the system can cope with the requests *)
+        Lwt_unix.sleep 0.2
+
+  let watch ~direct typ queue =
+    let root = cache_of typ in
+    let __FUN = __FUNCTION__ in
+    let rec loop () =
+      (* When the pushing side is disengaged it doesn't push events to the
+         queue, this means that trying to drain it completely would leave the
+         pulling side locked waiting when the queue is empty.
+           - Read the number of elements in the queue while draining it and
+             then switch to read the contents from the cache; or
+           - Switch immediately to reading the contents from cache and ignore
+             the contents of the queue by calling an specialized method in the
+             queue module to drain it.
+      *)
+      let get_action () =
+        let@ () = with_lock queue.lock in
+        match queue.state with
+        | Disengaged when Lwt_bounded_stream.size queue.queue < 1 ->
+            let* () = Lwt.pause () in
+            Lwt.return Update_all
+        | Direct ->
+            let* () = Lwt.pause () in
+            Lwt.return Wait
+        | _ -> (
+            let* elem = Lwt_bounded_stream.get queue.queue in
+            match elem with
+            | None ->
+                raise (Failure "Other side closed channel, cannot continue")
+            | Some elem ->
+                Lwt.return (File (elem, path_of_key root elem))
+          )
+      in
+      let* action = get_action () in
```

Comment:
To avoid race conditions it might be safer to acquire and hold the lock over both get_action and resolve to ensure queue state cannot change inbetween
```
let* () =
  let@ () = with_lock queue.lock in
  let* action = get_action () in
  resolve queue direct root action
in
loop ()
```